[DRAFT] Ami minimum age

cebernardi opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
chore:           <-- Smaller changes that impact behavior but aren't large enough to be features
perf:            <-- Code changes that improve performance but do not impact behavior
docs:            <-- Documentation changes that do not impact code
test:            <-- Test changes that do not impact behavior
ci:              <-- Changes that affect test or rollout automation
!${type}:        <-- Include ! if your change includes a backwards incompatible change.

Please review the Karpenter contribution docs at https://karpenter.sh/docs/contributing/ before submitting your pull request.
-->

Fixes #5382 <!-- issue number -->

**Description**
This is a draft PR that proposed a possible implementation for the introduction of a "ami minimum age" parameter.
I opened it as a draft to get early feedback. 

Not sure if this feature is already worked on by the maintainers or if there are other plans about it.

**How was this change tested?**
It's not exactly tested (a part for a very small subset of use cases, locally), all the unit tests are missing.

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates <!-- docs must be added to /preview to be included in future version releases -->
- [ ] Yes, issue opened: # <!-- issue number -->
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cebernardi commented:
Basically the proposed change is the following:

we query SSM for the recommended Amis, and then we check with the `ec2 describe-image` if the creation date complies with the newly introduced `amiminimumage` parameter. If not, we pick the next ami that matches the required minimum age. 

some assumptions here:
- if a recommended ami is more recent than the required `miMinumAge`, we need to retrieve anyway a "valid" image (is this even true? [This test suite](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/suite_test.go#L189) seems to suggest that in other scenarios we accept we don't have a valid ami)
- we can't query the SSM parameter history of the recommended amis (I have a very powerful role and I couldn't get the parameter history, so I assumed the parameter was protected)
- the "approved" eks node images are always published with the naming convention of `amazon-eks-arm64-node-<eks-version>-` or similar. It sounds fragile, I hope I'm missing something: the ec2 api doesn't allow to specify the image location in the filter, so I'm wondering if the name of the ami is robust enough.

cebernardi commented:
For brevity, the code only implements the proposed change for AL2 family, to gather some feedback about the strategy overall, so the other families are missing (on purpose)

njtran commented:
Hey @cebernardi can you go into a bit of your use case here? For features like these, it would be nice to talk through what you're trying to achieve and if this is the best way to go about it. Usually the best way to do this is to open a RFC (review for comments) or design doc that details the use-cases and the reason you picked this solution.

cebernardi commented:
I was picking up from [this discussion](https://github.com/aws/karpenter-provider-aws/issues/5382), the `amiMinimumAge` was proposed by @jonathan-innis , and it seemed perfect for the use case discussed on the issue.

I'm happy to open a RFC, but given that's Jonathan's idea, would you still suggest me to open one?

njtran commented:
I think it'd be best to open a RFC, as it seems like he was just proposing an idea, but there wasn't much discussion on if we want to do this.

cebernardi commented:
Sounds good!

cebernardi commented:
@njtran I reverted the code and added a design document. 
Please let me know if I interpreted your request correctly (I couldn't find any reference to RFC format on the documentation)

cebernardi commented:
should I also change it from draft to a proper PR?

njtran commented on designs/ami-minimum-age.md:
```diff
@@ -0,0 +1,70 @@
+  name: my-node-class
+spec:
+  amiFamily: AL2
+  amiMinimumAge: 2w
```
I feel like this might make more sense to be a sub-field of an AMI Selector Term. I could imagine users wanting to set a amiMinimumAge as another selection logic for each term, rather than prescribing it as an overlay over all selector terms.

cebernardi replied:
I see your point.

Just to double check, if `amiMinimumAge` was a sub-field of AMI Selector Term, it wouldn't be available to users that only want to set `amiFamily`.

Users that only need to specify  the `amiFamily` and a minimum age, would need to use a configuration like the following (assuming AL2 and eks 1.28):

```
amiSelectorTerms:
- name: "amazon-eks*node-1.28*"
   minimumAge: 5d
```
it seems to complicate a bit the simplest configuration scenario (I need the most recent AMI of the specified family, older than `minAge`), but I guess you see `amiMinimumAge` as a sort of "advanced" configuration option that warrants some more maintenance work from users? 
I'm thinking in particular at use cases when eks gets updated. Where before users that only specified `amiFamily` didn't have to upgrade their NodeClasses, now they would have to push a new NodeClasses configuration.

I'm wondering if we could do both?
The way the AMIs are discovered right now is mutually exclusive: https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/ami.go#L118

if there are `amiSelectorTerms` specified, the `amiFamily` is ignored. 

Can we adopt the same logic and have something like
```
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: my-node-class
spec:
  amiFamily: AL2
  amiMinimumAge: 2w
```

as well as
```
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: my-node-class
spec:
  amiSelectorTerms:
  - name: "amazon-eks*node-1.28*"
     minimumAge: 5d
```

and have the best of the 2 world?

njtran replied:
You bring up a good point, and even if there was a way to have an amiSelectorTerm, it's not _as_ simple to define it within each term. On the topic of having it in two places, I'd really opt against that, as that it may be a best of both worlds for some people, but is likely to have semantic and education confusion later.

cebernardi replied:
So we c/should keep it as first level attribute (`spec.amiMinimumAge`)?

njtran commented on designs/ami-minimum-age.md:
```diff
@@ -0,0 +1,70 @@
+
+Karpenter allows users to specify TTL for nodes, which configures the maximum age of nodes in the cluster. 
+
+This is useful to ensure that nodes are replaced periodically, and helps to ensure that the cluster is running on the latest AMIs available, 
+simplifying and automating the security patching process.
```
One thing i'm curious about is that you're looking to use this in response to the AMI issue that you've tagged below. What would you have set on this amiMinimumAge if it was specified?

One of the issues with this is if you have a new AMI released as a CVE patch, or a roll forward from a faulty image, would you still want to wait the amiMinimumAge to get that AMI? I think in almost all cases like that, you'd want to grab the image as soon as possible.

cebernardi replied:
I think we would pick an interval that will comply with the most stringent security SLA that our company has defined, something like 5d.

We sourced the referred bad AMI in our production clusters the same day it was published. If we had this mechanism we would have pulled it in only in our integration environment, where we would (hopefully) detected the issue, and stopped the production clusters from pulling the new AMI until the problem was solved.

In the scenarios you're describing (an urgent fix-forward AMI release) we would have released another version of the EC2NodeClass, maybe with no or a shorter minimum age. That would have triggered a drift and reclaimed the nodes out of compliance (please correct me if I'm wrong).

this is correlated with [this comment](https://github.com/aws/karpenter-provider-aws/pull/5551#discussion_r1481936955) about how a new AMI and drift disruption work. I think I might not have a clear picture of it.

njtran commented on designs/ami-minimum-age.md:
```diff
@@ -0,0 +1,70 @@
+
+## Background
+
+Karpenter allows users to specify TTL for nodes, which configures the maximum age of nodes in the cluster. 
```
I think you're referring to expiration here, which I don't think really has relevance here. In the case of an AMISelector Term that doesn't use AMI ID, we look periodically at if the AMI has changed every 5 minutes, and roll that out. This minimumAge would be a way to modify our discovery logic to change what that resolved AMI would be.

cebernardi replied:
my understanding is that a new AMI doesn't trigger a drift, so nodes running on the old AMI won't be recycled. 
So setting an expiration makes sure that the entire cluster is bounced to the new AMI, over a certain period of time.
Is this correct, or when a new AMI is detected, the nodes running the old AMI are marked as drifted?

cebernardi replied:
found this in the documentation: https://karpenter.sh/docs/concepts/disruption/#special-cases-on-drift but not sure if it's related also to configurations where no `amiSelectorTerms` is set?

njtran reviewed (commented):
Hey @cebernardi thanks for cutting it.  I've left a few high level comments.  You can take it out of draft if you'd like.

njtran commented:
Hey @cebernardi, planning on taking another look at this next week, but in the meantime, there's some discussion amongst the maintainers to figure out how to handle the issues that are motivating you to cut this. I'll post here when I have more information on this.

njtran commented:
Hey @cebernardi, I know I said I would give you an update on this last week, but still figuring some things out. We're currently figuring out if there are any additional guardrails that we can implement in clusters that would allow you to control the rate that new AMIs are rolled out in your cluster. Can you give some more details about how you detected the issue with the AMI you had? How would you have best liked to see the issue detected?

cebernardi commented:
Hi @njtran 
we detected the issue because we got a spike of applications failing to schedule on the affected nodes.
The spike was happening both in test and production, which was confusing because we couldn't immediately link it to any change we deployed and we were actively monitoring.

We're not testing _specifically_ new AMIs, but we have synthetic use-cases running twice a day in our test environments.

Ideally the spike would have been happening only in test environments, so the higher environments would have been spared.

Does it answer your question?

njtran commented:
@cebernardi Ah so you use automatic upgrades (no AMISelectorTerms) in both dev and production?

cebernardi commented:
Yes

> @cebernardi Ah so you use automatic upgrades (no AMISelectorTerms) in both dev and production?

Yes

njtran commented:
It may be worth your time to read through https://karpenter.sh/docs/tasks/amitasks/#choosing-ami-tasks. We added some more details on what we think is the best way to model your AMI rollouts. Not saying this fixes everything, but this should help a bit while we're still figuring out things on our end.

njtran commented:
Closing this due to a conversation that @jonathan-innis and @cebernardi had at KubeCon that we don't think minimum age is the right approach here.

cebernardi commented:
we tested the approach suggested [here](https://github.com/aws/karpenter-provider-aws/pull/5551#issuecomment-1979677824), and we found it adds an (avoidable, in my opinion) overhead on the operational side. 
Please see [this comment](https://github.com/aws/karpenter-provider-aws/issues/5382#issuecomment-2056555016).

Can we reopen this?
